ledger/internal/appcow.go

```diff
@@ -670,32 +677,47 @@ func applyStorageDelta(cb *roundCowState, addr basics.Address, aapp storagePtr,
 			if !exist {
 				return fmt.Errorf("could not find existing params for %v", aapp.aidx)
 			}
-			params = params.Clone()
+			paramsClone := params.Params.Clone()
+			params.Params = &paramsClone
 			if (storeDelta.action == allocAction && len(storeDelta.kvCow) > 0) ||
-				(storeDelta.action == remainAllocAction && params.GlobalState == nil) {
+				(storeDelta.action == remainAllocAction && params.Params.GlobalState == nil) {
 				// allocate KeyValue for
 				// 1) app creation and global write in the same app call
 				// 2) global state writing into empty global state
-				params.GlobalState = make(basics.TealKeyValue)
+				params.Params.GlobalState = make(basics.TealKeyValue)
 			}
 			// note: if this is an allocAction, there will be no
 			// DeleteActions below
 			for k, v := range storeDelta.kvCow {
 				if !v.newExists {
-					delete(params.GlobalState, k)
+					delete(params.Params.GlobalState, k)
 				} else {
-					params.GlobalState[k] = v.new
+					params.Params.GlobalState[k] = v.new
 				}
 			}
-			cb.mods.NewAccts.UpsertAppParams(addr, aapp.aidx, &params)
+			// fetch AppLocalState to store along with updated AppParams
+			var state ledgercore.AppLocalStateDelta
+			if st, exist, err := cb.lookupAppLocalState(addr, aapp.aidx); err != nil {
+				return fmt.Errorf("fetching storage (global=%v) failed for (%s, %d) AppLocalState: %w", aapp.global, addr.String(), aapp.aidx, err)
+			} else if exist {
+				state = st
+			}
+			cb.mods.NewAccts.UpsertAppResource(addr, aapp.aidx, params, state)
 		}
 	} else {
 		switch storeDelta.action {
 		case deallocAction:
 			if _, ok := cb.mods.NewAccts.GetData(addr); !ok {
 				return fmt.Errorf("dealloc consistency check (global=%v) failed for (%s, %d)", aapp.global, addr.String(), aapp.aidx)
 			}
-			cb.mods.NewAccts.UpsertAppLocalState(addr, aapp.aidx, nil)
+			// fetch AppParams to store along with deleted AppLocalState
+			var params ledgercore.AppParamsDelta
+			if ap, exist, err := cb.lookupAppParams(addr, aapp.aidx); err != nil {
+				return fmt.Errorf("fetching storage (global=%v) failed for (%s, %d) AppLocalState: %w", aapp.global, addr.String(), aapp.aidx, err)
+			} else if exist {
+				params = ap
+			}
+			cb.mods.NewAccts.UpsertAppResource(addr, aapp.aidx, params, ledgercore.AppLocalStateDelta{Deleted: true})
```

Comment:
maybe we need to test dealloc local state and then dealloc app params in two subsequent tnxs